### PR TITLE
[GSoC] Dividing system of ODEs into sub-systems

### PR DESCRIPTION
**Pull request by mijo2**
#### Brief description of what is fixed or changed
Completed `_component_division` private function. This function can now divide systems of ODEs into required sub-systems to solve the ODEs by dividing it.

#### Other comments
The library now can handle systems of ODEs such that a part of them is solvable and the other part of them can be solved if the first solvable part is solved. The example below demonstrates this:
```
In [39]: f, g, h, k = symbols('f g h k', cls=Function) 
    ...: eqs = [ 
    ...: Eq(f(x).diff(x), f(x)),  
    ...: Eq(g(x).diff(x), f(x)),  
    ...: Eq(h(x).diff(x), h(x)), 
    ...: Eq(k(x).diff(x), h(x)**4 + k(x))] 
    ...: dsolve(eqs) 
Out[39]: 
[Eq(f(x), C1*exp(x)),
 Eq(g(x), C1*exp(x) + C2),
 Eq(h(x), C3*exp(x)),
 Eq(k(x), (C4 + Integral(C3**4*exp(3*x), x))*exp(x))]

```
The above system as a whole doesn't have a solver but by first solving for `f(x)` and `g(x)` separately, and then just solving for `h(x)` and substituting that solution to solve for `k(x)` leads to the solution using just linear solvers and clever substitutions and division of system.
#### Release Notes

* solvers
  * Added component division technique to divide the system of ODEs into logical sub-systems and solving each of these separately.

### CODE REVIEW

**Comment by mijo2** [comment]
@oscarbenjamin I have added the component division logic in these two commits. But before proceeding, I just wanted to confirm something:

Now, in the second commit, I have made `dsolve_system` such that it will try to divide every system of ODEs that is passed. So, for a system given below:
```
    eq2 = (
        Eq(Derivative(z(t), t), k2*y(t)),
        Eq(Derivative(x(t), t), k3*y(t)),
        Eq(Derivative(y(t), t), (-k2 - k3)*y(t))
    )
```
Even though we can directly solve this system using our linear solvers, it is visible that we can divide the system of ODEs into sub-systems and solve them independently. So, with the current state of `dsolve_system`, the answer to the above system that will be returned is:
```
[Eq(z(t), C1 + Integral(C2*k2*exp(-k2*t - k3*t), t)),
 Eq(x(t), C3 + Integral(C2*k3*exp(-k2*t - k3*t), t)),
 Eq(y(t), C2*exp(t*(-k2 - k3)))]
```
compared to the previously obtained solution: 
```
[Eq(z(t), C1 - C2*k2*exp(t*(-k2 - k3))/(k2 + k3)),
 Eq(x(t), -C2*k3*exp(t*(-k2 - k3))/(k2 + k3) + C3),
 Eq(y(t), C2*exp(t*(-k2 - k3)))]
```
the current one has more number of unevaluated integrals. Should I proceed with the current strategy or should I just use the linear solvers without dividing the systems if the system as a whole can be solved by one of the solvers?

**Comment by oscarbenjamin** [comment]
If the system as a whole is homogeneous then splitting it into strongly connected components will lead to non-homogeneous subsystems. In the case of constant coefficients it's definitely better to avoid making the problem non-homogeneous so in that case we should solve the system as a whole. Otherwise if the system is nonlinear, not constant coefficient or non-homogeneous then it might always be preferable to split into strongly connected components.

If we had an undetermined coefficients solver for certain cases of constant coefficient non-homogeneous systems then using that while splitting the system into components might be better than solving the constant coefficient homogeneous system as a whole.

We should always split weakly connected components.

A tricky example would be this
```
dx/dt = y
dy/dt = x
dz/dt = x
dw/dt = z**2
```
Since w is nonlinear we can't use the linear solver for the system as a whole. The components are `{x, y}, {z}, {w}`. Ideally though we should solve `{x, y, z}` as a linear, constant coefficient, homogeneous subsystem rather than solve `{x, y}` and then solve a non-homogeneous system for `z`.

**Comment by mijo2** [comment]
So, we can form the graph in such a way that it will think that if a system is linear constant coeff homogeneous, then it would be a part of a strongly connected component. I would build a separate function for building graph to find strongly connected components to better implement this approach.

File: sympy/solvers/ode/systems.py [hunk 1]
```diff
@@ -1076,11 +1076,78 @@ def _preprocess_eqs(eqs):
...
+
+            if not _is_type1((check_eqs, funcs), t):
+                break
+
+            old_eqs = sccs[0][0]
+            sccs[0] = (old_eqs + eqs, old_funcs + funcs)
+            end_idx += 1
+
+        del sccs[1:end_idx]
+
```
Comment (mijo2) [inline sympy/solvers/ode/systems.py:1124]: @oscarbenjamin In this function, I have added a way to combine the sub-systems when they form a type 1(constant coeff homogeneous) system of ODEs. 

So, the first thing to be observed is that if the first sub-system is not of Type 1, then the other sub-systems can't be of Type 1 either. Hence, if there is exists a larger Type 1 sub-system, then it starts from the first component in a list of strongly connected components. Now, to check if the second, third, ..., etc systems are Type 1, we have to replace the dependent variables from the previous components as one of the dependent variables of this component. This makes the check a whole lot easier. If the current component is Type 1, we simply add this component with the first component and keep doing it until we reach a sub-system which is not of Type 1.

Is this logic correct?

Comment (oscarbenjamin) [reply sympy/solvers/ode/systems.py:1122]: How does it handle a case like the following?
```
dx/dt = x**2
dy/dt = z
dz/dt = y
dw/dt = x + z
```
Here the components are `{x}, {y, z}, {w}` but you might get them in that order or you might get them in the order `{y, z}, {x}, {w}`. The component `{y, z}` is linear with constant coefficients but it might come before or after `{x}` which is nonlinear.

Comment (mijo2) [reply sympy/solvers/ode/systems.py:1122]: Well then it would certainly be a problem. I think to avoid this, what can be done is checking the graph to see if one connected component has a directed edge to another. I would try to implement it soon.

Comment (mijo2) [reply sympy/solvers/ode/systems.py:1150]: I have tried to fix this in the recent commit. In that, I have built a graph for strongly connected components in the form of `parents` list. This is a list of lists, each index represents the strongly connected component(scc). Then in `_combine_systems` function, I combine the systems which are of type1 and along with that keep track of its parent sccs too. If any parent scc is not type1, then this scc won't be combined. After the check, we combine the scc to the scc with the lowest index in the list of sccs to keep the topological order intact.

Comment (oscarbenjamin) [reply sympy/solvers/ode/systems.py:1150]: That sounds reasonable.

**Comment by mijo2** [comment]
@oscarbenjamin Except the type 1 combination, should we just try to identify if the whole system is solvable by a solver(like how it was before this PR) since if we don't do this, then lots of test cases may have nested integrals. If it's fine to have such expressions, then I would start working on replacing the tests with the new values along with extensively testing the component division.

**Comment by oscarbenjamin** [comment]
Can you give an example system to explain what you mean?

**Comment by mijo2** [comment]
```
In [6]: f, g, h = symbols('f g h', cls=Function) 
   ...: eqs = [ 
   ...: Eq(f(x).diff(x), 2*f(x)),  
   ...: Eq(g(x).diff(x), f(x) + 1),  
   ...: Eq(h(x).diff(x), g(x))] 
   ...: dsolve(eqs)                                                                                                                                                                                         
Out[6]: 
[Eq(f(x), C1*exp(2*x)),
 Eq(g(x), C2 + Integral(C1*exp(2*x) + 1, x)),
 Eq(h(x), C3 + Integral(C2 + Integral(C1*exp(2*x) + 1, x), x))]
```

**Comment by oscarbenjamin** [comment]
Okay that makes sense. Yes, maybe if the whole system matches one of the linear types we should just use that solver.

**Comment by mijo2** [comment]
@oscarbenjamin I am adding this check in the beginning of the weakly connected components solver. I am still making it mandatory for solving weakly connected components separately as it makes sense to do so.

**Comment by mijo2** [comment]
@oscarbenjamin I have encountered a problem while testing a solution. So, our `dsolve_system` method can now handle many new types of systems of ODEs. One such system is:

```
In [6]:     eqs1 = [Eq(Derivative(f(x), x), 2*f(x)), 
   ...:             Eq(Derivative(g(x), x), f(x)), 
   ...:             Eq(Derivative(h(x), x), h(x)), 
   ...:             Eq(Derivative(k(x), x), h(x)**4 + k(x))] 
```

Solution for this is:

```
In [14]: dsolve(eqs1)                                                                                                                                                                                       
Out[14]: 
[Eq(f(x), C1*exp(x)),
 Eq(g(x), 2*C1*exp(x) + C2),
 Eq(h(x), C3*exp(x)),
 Eq(k(x), (C4 + Integral(C3**4*exp(3*x), x))*exp(x))]
```

 But, when I try to check the validity of the solution, I get an error like below:

```
In [15]: checksysodesol(eqs1, dsolve(eqs1))                                                                                                                                                                 
Out[15]: (False, [-C1*exp(x), C1*exp(x), 0, 0])
```

But, if I try to do the below, it works:

```
In [16]: checksysodesol(eqs1[:2], dsolve(eqs1[:2]))                                                                                                                                                         
Out[16]: (True, [0, 0])
In [17]: checksysodesol(eqs1[2:], dsolve(eqs1[2:]))
Out[17]: (True, [0, 0])

```
Should I create an issue or try to fix `checksysodesol`? For now, I am adding test cases by splitting them up.

**Comment by mijo2** [comment]
@oscarbenjamin The main part of this PR is done. You can have a look.

**Comment by oscarbenjamin** [comment]
I find the code here quite hard to follow so I wrote a few functions myself to split the system into connected components:
```python
def eqs2dict(eqs, funcs):
    eqsorig = {}
    eqsmap = {}
    funcset = set(funcs)
    for eq in eqs:
        f1, = eq.lhs.atoms(AppliedUndef)
        f2s = (eq.rhs.atoms(AppliedUndef) - {f1}) & funcset
        eqsmap[f1] = f2s
        eqsorig[f1] = eq
    return eqsmap, eqsorig


def dict2graph(d):
    nodes = list(d)
    edges = [(f1, f2) for f1, f2s in d.items() for f2 in f2s]
    G = (nodes, edges)
    return G


def _split_ode_system(eqs, funcs, t):
    eqsmap, eqsorig = eqs2dict(eqs, funcs)

    subsystems = []
    for cc in connected_components(dict2graph(eqsmap)):
        eqsmap_c = {f: eqsmap[f] for f in cc}
        sccs = strongly_connected_components(dict2graph(eqsmap_c))
        subsystems.append([[eqsorig[f] for f in scc] for scc in sccs])

    return subsystems
```
Also here is a function that can combine the type1 subsystems:
```python
def _combine_type1_subsystems(subsystem, funcs, t):
    indices = [i for i, sys in enumerate(subsystem) if _is_type1((sys, funcs), t)]
    remove = set()
    for ip, i in enumerate(indices):
        eqsmap, eqs_orig = eqs2dict(subsystem[i], funcs)
        funcs_i = list(eqsmap)
        for j in indices[ip+1:]:
            if any(eq2.has(*funcs_i) for eq2 in subsystem[j]):
                subsystem[j] = subsystem[i] + subsystem[j]
                remove.add(i)
    subsystem = [sys for i, sys in enumerate(subsystem) if i not in remove]
    return subsystem
```
Do these do the same as the code in the PR? If not what exactly is the difference?

**Comment by mijo2** [comment]
The code you wrote does the same thing as what I wrote but is a lot simpler. I will use this in the codebase and check if it works properly or not.

**Comment by mijo2** [comment]
@oscarbenjamin I have used your code with a few minor changes. Thanks for the simplification. I would try to write simpler code from now on in future PRs with the best of my capabilities.

File: sympy/solvers/ode/tests/test_systems.py [hunk 2]
```diff
@@ -1210,6 +1210,81 @@ def test_sysode_linear_neq_order1_type4():
     assert checksysodesol(eqs8, sol8) == (True, [0, 0, 0, 0])
 
 
+def test_component_division():
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/tests/test_systems.py:1213]: There should be direct tests for the `_component_division` function and the `eqs2dict` and `dict2graph` functions.

File: sympy/solvers/ode/tests/test_systems.py [hunk 3]
```diff
@@ -1761,3 +1836,25 @@ def test_dsolve_system():
...
+    f, g = symbols('f g', cls=Function)
+    x, y = symbols('x y')
+
+    eqs = [f(x).diff(x, 2), g(x).diff(x)]
+    with raises(ValueError):
+        dsolve(eqs) # NotImplementedError would be better
+
+    eqs = [f(x).diff(x) - x, f(x).diff(x) + x]
+    with raises(ValueError):
+        # Could also be NotImplementedError. f(x)=0 is a solution...
```
Comment (oscarbenjamin) [inline sympy/solvers/ode/tests/test_systems.py:1851]: Here `f(x) = 0` is not a solution.

Comment (mijo2) [reply sympy/solvers/ode/tests/test_systems.py:1851]: I would remove that comment.

**Comment by oscarbenjamin** [comment]
This looks good. Good work!

I'm going to merge this now but then I'm going away for a week or so. I'll be able to look at your next PRs when I get back.

**Comment by mijo2** [comment]
Ok, I would start working on Higher order to first order reduction next. You can check it out when you are free.
